File: csi/java/src/main/java/org/shypl/biser/csi/server/OutgoingMessages.java
package org.shypl.biser.csi.server;

import java.util.Collection;
import java.util.Deque;
import java.util.LinkedList;

public class OutgoingMessages {
	
	private int                    idCounter = 0;
	private Deque<OutgoingMessage> list      = new LinkedList<>();
	private int lastReleasedId;
	
	public OutgoingMessage create(byte[] data) {
		OutgoingMessage message = new OutgoingMessage(nextId(), data);
		list.addLast(message);
		return message;
	}
	
	public void releaseFirst() {
		lastReleasedId = list.removeFirst().id;
	}
	
	public void releaseTo(int id) {
		while (lastReleasedId != id && !list.isEmpty()) {
			releaseFirst();
		}
	}
	
	public Collection<OutgoingMessage> getQueue() {
		return list;
	}
	
	private int nextId() {
		int id = ++idCounter;
		if (id == 0) {
			id = ++idCounter;
		}
		return id;
	}
	
}

File: kotlin/biser-kotlin-js/src/main/kotlin/org/shypl/biser/csi/client/OutgoingMessages.kt
package org.shypl.biser.csi.client

internal class OutgoingMessages {
	val queue: List<OutgoingMessage>
		get() = list
	
	private val list = mutableListOf<OutgoingMessage>()
	private var idCounter = 0
	private var lastReleasedId = 0
	
	fun create(data: ByteArray): OutgoingMessage {
		val message: OutgoingMessage = OutgoingMessage(nextId(), data)
		list.add(message)
		return message
	}
	
	fun releaseTo(id: Int) {
		while (lastReleasedId != id && !list.isEmpty()) {
			val first: OutgoingMessage = list.removeAt(0)
			lastReleasedId = first.id
		}
	}
	
	
	private fun nextId(): Int {
		var id = ++idCounter
		if (id == 0) {
			id = ++idCounter
		}
		return id
	}
}
